### DISCUSSION TITLE
NextUrl Pagination - Return data until response = empty

### DISCUSSION THREAD

**Discussion by Mkallday** [thunderclient/thunder-client-support#1470] [Q&A]
### First Request 
1.Tests Tab: **Set Env Variable** for _nextUrl_
![image](https://github.com/rangav/thunder-client-support/assets/36720153/f030a0c0-c8ef-41d0-a56f-06c318b8a9c8)
### Second Request 
1. Tests Tab: **Set Env Variable** for _nextUrl_
2. Pre Run Tab: Run the **First Request**
![image](https://github.com/rangav/thunder-client-support/assets/36720153/272e6e61-3626-4ca1-a4ef-b22d1c759128)
OR
### Pre Run Tab: Both requests
![image](https://github.com/rangav/thunder-client-support/assets/36720153/115c941c-8b12-4108-a65b-95e2bfb6d71f)

**Comment by rangav** [comment]
@Mkallday no, you dont need 2 request for pagination to work.

- You can use the Tests Tab -> scripting and use custom script then you can run the same request until a desired result.
- See example here - https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#retry-request

**Reply by Mkallday** [reply]
Can you be a little more specific? I've been arguing with ChatGPT all day and I have yet to successfully paginate.

**Reply by Mkallday** [reply]
@rangav Could you please give me a little more direction? Claude, Cody, ChatGPT and Genie aren't producing correct responses.   I'm using the paid version and I won't ask anymore specific questions this year

**Reply by rangav** [reply]
no problem you can email us regarding subscription confirmation, better not to share here for privacy/security.

Can you please share your account number or email used in below contact form
https://www.thunderclient.com/contact

**Reply by Mkallday** [reply]
Just did so! @rangav

**Comment by rangav** [comment]
- remove all the pre requests in Pre Run tab
- remove the existing Tests also
- You can use custom scripting to get this functionality and use `Tests tab -> scripting`

Modify the below code as per your requirements

```js
let json = tc.response.json;
let nextUrl = json["@attributes"].next;

if(nextUrl)
{
    tc.setVar('next_url', nextUrl)
    console.log("running next request", nextUrl);

    await tc.retryRequest();
}
else
{
   console.log("got all the data");
}
```

**Reply by rangav** [reply]
if the above `await tc.retryRequest();` does not work.

you can also use `await tc.runRequest(tc.request.id);`

see the difference between the two functions here
https://github.com/rangav/thunder-client-support/blob/master/docs/tc-types.d.ts#L96

**Reply by Ak-sky** [reply]
I tried but `await tc.retryRequest();` kept on calling the first request in loop instead of calling the nextUrl

**Reply by Ak-sky** [reply]
and if I am sending custom requests using axios

```
let nextUrl = JSON.stringify(jsonData['next'])

if(nextUrl)
{
    tc.setVar('next_url', nextUrl)
    console.log("running next request", nextUrl);
    const response = await axios.get(nextUrl);
}
else
{
  console.log("got all the data");
}
```

I am getting below error

`Error in Script: - connect ECONNREFUSED 127.0.0.1:80 `

Am I supposed to send token with axios? If yes how?

**Reply by Ak-sky** [reply]
Can someone please reply on this?

**Reply by rangav** [reply]
We will schedule call to resolve this

please share your availability using below form
https://www.thunderclient.com/contact

**Reply by Mkallday** [reply]
> * remove all the pre requests in Pre Run tab
> * remove the existing Tests also
> * You can use custom scripting to get this functionality and use `Tests tab -> scripting`
> 
> Modify the below code as per your requirements
> 
> ```js
> let json = tc.response.json;
> let nextUrl = json["@attributes"].next;
> 
> if(nextUrl)
> {
>     tc.setVar('next_url', nextUrl)
>     console.log("running next request", nextUrl);
> 
>     await tc.retryRequest();
> }
> else
> {
>    console.log("got all the data");
> }
> ```

I STILL CAN'T PAGINATE. 
![image](https://github.com/user-attachments/assets/18819ccb-005b-471c-8810-a5c63d6ae97d)


- Previously, when I use your code above, it would "run" forever and eventually time out. 
- Today, I watched the Terminal Output of this API call and this script running the same URL. It's not getting the NEXTURL. 
- Claude + Cline have made dozens of variations, but NOTHING IS WORKING.

**Comment by Mkallday** [comment]
@rangav

**Reply by rangav** [reply]
I will verify and get back to you.
